examples/Azure_IoT_Adu_ESP32/Azure_IoT_Adu_ESP32.ino

```diff
@@ -0,0 +1,1054 @@
+  mqtt_config.client_cert_pem = IOT_CONFIG_DEVICE_CERT;
+  mqtt_config.client_key_pem = IOT_CONFIG_DEVICE_CERT_PRIVATE_KEY;
+#else // Using SAS key
+  mqtt_config.password = (const char*)az_span_ptr(sasToken.Get());
+#endif
+
+  mqtt_config.keepalive = 30;
+  mqtt_config.disable_clean_session = 0;
+  mqtt_config.disable_auto_reconnect = false;
+  mqtt_config.event_handle = mqtt_event_handler;
+  mqtt_config.user_context = NULL;
+  mqtt_config.cert_pem = (const char*)ca_pem;
+  mqtt_config.buffer_size = 2048;
+
+  mqtt_client = esp_mqtt_client_init(&mqtt_config);
+
+  if (mqtt_client == NULL)
+  {
+    Logger.Error("Failed creating mqtt client");
+    return 1;
+  }
+
+  esp_err_t start_result = esp_mqtt_client_start(mqtt_client);
+
+  if (start_result != ESP_OK)
+  {
+    Logger.Error("Could not start mqtt client; error code:" + start_result);
+    return 1;
+  }
+  else
+  {
+    Logger.Info("MQTT client started");
+    return 0;
+  }
+}
+
+/*
+ * @brief           Gets the number of seconds since UNIX epoch until now.
+ * @return uint32_t Number of seconds.
+ */
+static uint32_t get_epoch_time_in_seconds() { return (uint32_t)time(NULL); }
+
+static void establish_connection()
+{
+  connect_to_wifi();
+  initialize_time();
+  initialize_iot_hub_client();
+  (void)initialize_mqtt_client();
+}
+
+static void get_telemetry_payload(az_span payload, az_span* out_payload)
+{
+  az_result rc;
+  az_span original_payload = payload;
+
+  payload = az_span_copy(payload, AZ_SPAN_FROM_STR("{ \"msgCount\": "));
+  rc = az_span_u32toa(payload, telemetry_send_count++, &payload);
+  (void)rc;
+  payload = az_span_copy(payload, AZ_SPAN_FROM_STR(" }"));
+  payload = az_span_copy_u8(payload, '\0');
+
+  *out_payload = az_span_slice(
+      original_payload, 0, az_span_size(original_payload) - az_span_size(payload) - 1);
+}
+
+static void send_telemetry()
+{
+  az_span telemetry = AZ_SPAN_FROM_BUFFER(telemetry_payload);
+
+  Logger.Info("Sending telemetry ...");
+
+  // The topic could be obtained just once during setup,
+  // however if properties are used the topic need to be generated again to
+  // reflect the current values of the properties.
+  if (az_result_failed(az_iot_hub_client_telemetry_get_publish_topic(
+          &hub_client, NULL, telemetry_topic, sizeof(telemetry_topic), NULL)))
+  {
+    Logger.Error("Failed az_iot_hub_client_telemetry_get_publish_topic");
+    return;
+  }
+
+  get_telemetry_payload(telemetry, &telemetry);
+
+  if (esp_mqtt_client_publish(
+          mqtt_client,
+          telemetry_topic,
+          (const char*)az_span_ptr(telemetry),
+          az_span_size(telemetry),
+          MQTT_QOS1,
+          DO_NOT_RETAIN_MSG)
+      < 0)
+  {
+    Logger.Error("Failed publishing");
+  }
+  else
+  {
+    Logger.Info("Message published successfully");
+  }
+}
+
+// Arduino setup and loop main functions.
+
+void setup() { establish_connection(); }
+
+bool send_init_state = true;
+
+void loop()
+{
+  az_result result;
```

Comment:
loop() became a little more bloated than most Arduino users prefer. Maybe a point for improvement in the future.